src/types.ts

```diff
@@ -0,0 +1,111 @@
+export interface DOMAPI {
+  createFragment: () => DocumentFragment;
+  createElement: (tagName: string) => HTMLElement;
+  createElementNS: (namespaceURI: string, qualifiedName: string) => Element;
+  createTextNode: (text: string) => Text;
+  createComment: (text: string) => Comment;
+  insertBefore: (parentNode: Node, newNode: Node, referenceNode: Node | null) => void;
+  removeChild: (node: Node, child: Node) => void;
+  appendChild: (node: Node, child: Node) => void;
+  parentNode: (node: Node) => Node | null;
+  nextSibling: (node: Node) => Node | null;
+  setTextContent: (node: Node, text: string) => void;
+}
+
+import {AttachData} from './helpers/attachto'
+import {VNodeStyle} from './modules/style'
+import {On} from './modules/eventlisteners'
+import {Attrs} from './modules/attributes'
+import {Classes} from './modules/class'
+import {Props} from './modules/props'
+import {Dataset} from './modules/dataset'
+import {Hero} from './modules/hero'
+
+export type Key = string | number;
+
+export type VNodes = Array<VNode | null>;
+
+export interface VNode {
+  nt: number; // matches DOM Spec for Node's nodeType
```

Comment:
this is one of the big changes. `nt` stands for `nodeType`, which is a numeric value matching the DOM spec to speed up the process of diffing by not having to compare the `sel` string values anymore.

additionally, developers can specify their own comparison mechanism if they have specific constrains around the identity of ever vnode.